### PR TITLE
Add shell completion supported by clap

### LINKED ISSUE #18: Bash completion
Add bash auto completion that parses the markdown file and provides a list of commands (headings) that can be executed
Related to #17

### PR DESCRIPTION
**Pull request by DrSensor**
### Which issue does this fix?
Closes #18 

### Describe the solution
Inspired from starship [[1]][starship], this PR will add `init` subcommand to generate shell completion via `clap` crate [[2]][gen_completion_to].

[1] https://starship.rs/#quick-install (jump to step 2)
[2] https://docs.rs/clap/2.33.0/clap/struct.App.html#examples-48

[starship]: https://starship.rs/#quick-install
[gen_completion_to]: https://docs.rs/clap/2.33.0/clap/struct.App.html#method.gen_completions_to

### How to test
```bash
hub pr checkout 42
cargo install --path .

# if using bash
eval "$(mask init bash)"
```

As for me, I'm using fish shell and it works
```fish
mask init fish | source
```
Peek 2019-11-17 07-33

### Types of changes

- [x] Bug fix               
- [x] New feature           
- [ ] Breaking change

### CODE REVIEW

**Comment by jakedeichert** [comment]
Hey @DrSensor thanks for the PR! I'll review this week as I have some time off 👍 .

File: src/main.rs [hunk 1]
```diff
@@ -84,6 +92,36 @@ fn custom_maskfile_path_arg<'a, 'b>() -> Arg<'a, 'b> {
...
 
+fn shell_completion_subcommand<'a, 'b>(cli_app: App<'a, 'b>) -> App<'a, 'b> {
+    cli_app.subcommand(
+        SubCommand::with_name("init")
+            .about("Prints the shell function used to execute `mask`")
+            .arg(
+                &Arg::with_name("shell")
+                    .value_name("SHELL")
+                    .help("The name of the currently running shell")
+                    .possible_values(&["bash", "zsh", "fish", "elvish", "powershell"])
```
Comment (jakedeichert) [inline src/main.rs]: Just curious, are these all the possible values clap supports?

**Comment by jakedeichert** [comment]
Thanks again for this PR! It's awesome that clap has support for completions built-in.

So i've been thinking about the `init` command. I really want to avoid creating subcommands for mask that could conflict with a user's commands.

I was thinking, maybe we do this `mask --completion bash|fish|etc` instead? Using an option instead of a command aligns more with another existing pattern we have (`mask --maskfile path`).

Here's how the `--maskfile` is added https://github.com/jakedeichert/mask/blob/master/src/main.rs#L74-L85


My thinking here is that root level options `mask --whatever` are ideal for mask-specific logic. If we can stick to this pattern, that would be ideal.

**Comment by mrmurphy** [comment]
Wow, would love to see this get merged <3 @DrSensor or @jakedeichert got a moment to take a look? Wish I knew any Rust so I could do it myself :D

**Comment by DrSensor** [comment]
Ah, thanks @mrmurphy for bringing this up.
I agree with the `--completion` flag and I think it should render all commands invalid.
Summary, it should error when user do `mask --completion=bash command`.

I just realized there is one problem with this feature. It can't be added in the startup script ( e.g _.config.fish_ or _.bash_profile_ ).
User can only register the completion manually on the **same dir** as _maskfile.md_ ( in my case `eval (mask --completion fish)` ).

File: src/main.rs (hunk 1 above)
Comment (DrSensor) [reply src/main.rs]: yups
https://docs.rs/clap/2.33.0/clap/enum.Shell.html

**Comment by mrmurphy** [comment]
Ah okay, that would be a problem!

**Comment by jakedeichert** [comment]
> I just realized there is one problem with this feature. It can't be added in the startup script ( e.g .config.fish or .bash_profile ).
>
> User can only register the completion manually on the same dir as maskfile.md ( in my case eval (mask --completion fish) ).

Hmmm... wouldn't a user be able to do `mask --maskfile ~/global.md --completion bash` and pop this into their `.bash_profile`?

**Comment by DrSensor** [comment]
Global maskfile should work but not with the local one.
(I wonder how version/workspace manager like [pyenv](https://github.com/pyenv/pyenv), [nvm](https://github.com/nvm-sh/nvm), [asdf](https://asdf-vm.com) can detect and auto-switch based on the local config 🤔)

**Comment by jakedeichert** [comment]
Well i guess i was assuming the user would run `mask --completion bash | source` once inside their project/local directory and then their current shell would have the proper completions available. It would have to be a manual step.

Though, I myself would probably make a helper function or alias in my bash_profile that would wrap `mask` to auto run this depending on which directory i'm in. But this part is definitely not mask's responsibility... user's should decide what's the best behavior for them, depending on which shell they use.

In the original issue, [jgroom posted a bash completions example](https://github.com/jakedeichert/mask/issues/18#issuecomment-514409413) which involved no changes to mask. I'm now wondering if we should just rely on community-supplied scripts that do simple parsing of the markdown headers for command names. I'd be very happy to link to these scripts (fish, bash, zsh, whatever) from the repo's readme.

**Comment by DrSensor** [comment]
I actually in favor of community-supplied scripts though it has the downside that the user need to make sure some built-ins CLI like _sed/grep_ are indeed _sed/grep_ (or at least being aliased properly).
But after fiddling around, I've found a way to generate completion script selectively. This allows
```console
mask --maskfile ~/global.md --completion bash subcmd subsubcmd ...
```
which generates the completion-script on certain sub/subsub...command. Here I can see a use case when the user needs different auto-completion when opening certain IDE.

Well, supporting both clap and community-supplied scripts is also an option :smile:

**Comment by DrSensor** [comment]
I choose option 2 then. It seems quite reasonable to wait for clap v3. In my last changes, it's quite weird that I need to call `build_subcommands` twice/3-times. Hope clap v3 open the possibility for refactoring mask so that it's easy for adding additional features 🥺

File: src/main.rs [hunk 2]
```diff
@@ -11,9 +11,9 @@ fn main() {
     let cli_app = App::new(crate_name!())
         .setting(AppSettings::VersionlessSubcommands)
         .setting(AppSettings::AllowNegativeNumbers)
-        .setting(AppSettings::SubcommandRequired)
```
Comment (DrSensor) [inline src/main.rs]: Side note for future references.
> Even without _subcommand filtering_, `AppSettings::SubcommandRequired` still need to be disabled or `unset_setting`

**Pull request closed without merging**